airlock/renderers.py

```diff
@@ -0,0 +1,119 @@
+import csv
+from dataclasses import dataclass
+from email.utils import formatdate
+from functools import cached_property
+from pathlib import Path
+
+from django.http import FileResponse
+from django.template import loader
+from django.template.response import SimpleTemplateResponse
+
+from airlock.business_logic import RequestFile
+
+
+@dataclass
+class Renderer:
+    template = None
+
+    abspath: Path
+    request_file: RequestFile = None
+
+    def get_response(self):
+        if self.template:
+            context = self.context()
+            response = SimpleTemplateResponse(self.template, context)
+        else:
+            response = FileResponse(self.abspath.open("rb"), filename=self.relpath.name)
+
+        for k, v in self.headers().items():
+            response.headers[k] = v
+
+        return response
+
+    def context(self):
+        raise NotImplementedError()
+
+    @cached_property
+    def template_path(self):
+        django_template = loader.get_template(self.template)
+        return Path(django_template.template.origin.name)
+
+    @cached_property
+    def stat(self):
+        return self.abspath.stat()
+
+    @cached_property
+    def cache_id(self):
+        if self.request_file:
+            cache_id = self.request_file.file_id
+        else:
+            cache_id = filesystem_key(self.stat)
+
+        if self.template:
+            cache_id += "-" + filesystem_key(self.template_path.stat())
```

Comment:
Given that this is cached, I'd be tempted to make this the hash of the template rather than its mtime. The hashing will only need to be done once per template per process, and given template files are relatively small anyway the overhead should be minimal. I recently discovered that `hashlib` has learned a new `file_digest` method which does the as efficiently as possible.

I'm always a little bit nervous of mtimes which come out of git/docker build chains as I've seen some odd things happen.

Reply:
Hmm, git mtimes is a thing, good point. Although that's really only going to be a thing in dev, were a bit of unnecessary cache invalidation is probably ok?

I went with mtime as I wanted it to invalidate caches in dev on when we edit templates (as I don't think django will auto-reload on template changes?)

But we could use stat() mtime to invalidate the cached file_digest?  That way if git changes mtimes, we'll re-calculate the hash, but not invalidate the HTTP cache?

Reply:
I'm going to punt on this for now. 

The files we serve and cache locally in dev are *not* in git and should be unmodified by git commands.  Also, it fails open, so I doubt we'd even notice.

We can revisit later if there's a problem.

Reply:
I'm dumb, the templates are affected by git of course.